wolfcrypt/src/port/Espressif/esp32_sha.c

```diff
@@ -0,0 +1,437 @@
+    word32 SHA_BUSY_REG = SHA_1_BUSY_REG;
+
+    ESP_LOGV(TAG, "enter esp_digest_state");
+
+    /* sanity check */
+    if(sha_type == SHA_INVALID) {
+        ESP_LOGE(TAG, "unexpected error. sha_type is invalid.");
+        return;
+    }
+    
+    SHA_LOAD_REG += (sha_type << 4);
+    SHA_BUSY_REG += (sha_type << 4);
+
+    if(ctx->isfirstblock == 1){
+        /* no hardware use yet. Nothing to do yet */
+        return ;
+    }
+
+    {
+        /* wait until idle */
+        esp_wait_until_idle();
+
+        /* LOAD final digest */
+        DPORT_REG_WRITE(SHA_LOAD_REG, 1);
+        /* wait until done */
+        while(DPORT_REG_READ(SHA_BUSY_REG) == 1){ }
+
+        esp_dport_access_read_buffer((word32*)(hash), SHA_TEXT_BASE,
+                                 esp_sha_digest_size(sha_type)/sizeof(word32));
+
+#if defined(WOLFSSL_SHA512) || defined(WOLFSSL_SHA384)
+        if(sha_type==SHA2_384||sha_type==SHA2_512) {
+            word32  i;
+            word32* pwrd1 = (word32*)(hash);
+            /* swap value */
+            for(i = 0; i <WC_SHA512_DIGEST_SIZE/4; i+=2 ) {
+                pwrd1[i]  ^= pwrd1[i+1];
+                pwrd1[i+1]^= pwrd1[i];
+                pwrd1[i]  ^= pwrd1[i+1];
+            }
+        }
+#endif
+    }
+
+    ESP_LOGV(TAG, "leave esp_digest_state");
+}
+
+#ifndef NO_SHA
+/*
+* sha1 process
+*/
+int esp_sha_process(struct wc_Sha* sha)
+{
+    int ret = 0;
+
+    ESP_LOGV(TAG, "enter esp_sha_process");
+    
+    word32 SHA_START_REG = SHA_1_START_REG;
+    
+    esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                        WC_SHA_BLOCK_SIZE);
+
+    ESP_LOGV(TAG, "leave esp_sha_process");
+    return ret;
+}
+/*
+* retrieve sha1 digest
+*/
+int esp_sha_digest_process(struct wc_Sha* sha, byte blockproc)
+{
+    int ret = 0;
+    
+    ESP_LOGV(TAG, "enter esp_sha_digest_process");
+
+    if(blockproc) {
+        word32 SHA_START_REG = SHA_1_START_REG;
+
+        esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                            WC_SHA_BLOCK_SIZE);
+    }
+
+    esp_digest_state(&sha->ctx, (byte*)sha->digest, SHA1);
+
+    ESP_LOGV(TAG, "leave esp_sha_digest_process");
+
+    return ret;
+}
+#endif /* NO_SHA */
+
+
+#ifndef NO_WOLFSSL_SHA256
+/*
+* sha256 process
+*/
+int esp_sha256_process(struct wc_Sha256* sha)
+{
+    int ret = 0;
+    
+    ESP_LOGV(TAG, "enter esp_sha256_process");
+    
+    word32 SHA_START_REG = SHA_1_START_REG + 16;
+
+    esp_process_block(&sha->ctx, SHA_START_REG, sha->buffer,
+                                            WC_SHA256_BLOCK_SIZE);
+
+    ESP_LOGV(TAG, "leave esp_sha256_process");
+
+    return ret;
+}
+/*
+* retrieve sha256 digest
+*/
+int esp_sha256_digest_process(struct wc_Sha256* sha, byte blockproc)
+{
+    int ret = 0;
+    
+    ESP_LOGV(TAG, "enter esp_sha256_digest_process");
+
+    if(blockproc) {
+        word32 SHA_START_REG = SHA_1_START_REG + SHA2_256 * 0x10;
```

Comment:
It looks like `SHA2_256` = 1, so this ends up being `SHA_1_START_REG + 16` just like above at 336. Perhaps it would be helpful to document these offsets or at least make them consistent. Is there a reason this won't work for all cases?

```
enum SHA_TYPE sha_type = sha->ctx.sha_type;
word32 SHA_START_REG = SHA_1_START_REG;  
SHA_START_REG += (sha_type << 4);
```